File: iosApp/iosApp/ui/historyDetail/HistoryDetailViewModel.swift
import Foundation
import SwiftUI
import shared

extension HistoryDetailScreen {
    @MainActor class HistoryDetailViewModel : ObservableObject, ViewLifeCycleObserver {

        private let historyId: String
        private var commonViewModel: HistoryDetailCommonViewModel?

        @Published private(set) var historyLoadStatus: LoadStatus<History>? = nil
        @Published private(set) var editingHistory: History? = nil
        @Published private(set) var showingElements: [HistoryElement]? = nil

        init(historyId: String) {
            self.historyId = historyId
        }

        func enableEditMode() {
            commonViewModel?.setEditMode()
        }

        func cancelEdit() {
            commonViewModel?.cancelEdit()
        }

        func editElement(newElement: HistoryElement, base64Data: String?) {
            commonViewModel?.editElement(newElement: newElement, imageBase64Data: base64Data)
        }

        func editTitle(newTitle: String) {
            commonViewModel?.editTitle(newTitle: newTitle)
        }

        func editDates(newDateRange: LocalDateRange) {
            commonViewModel?.editDates(newDateRange: newDateRange)
        }

        func createTextElement(text: String) {
            commonViewModel?.createTextElement(text: text)
        }

        func createImageElement(base64Data: String) {
            commonViewModel?.createImageElementFromBase64(base64Data: base64Data)
        }

        func swapElements(fromId: String, toId: String) {
            commonViewModel?.swapElements(fromId: fromId, toId: toId)
        }

        func deleteElement(element: HistoryElement) {
            commonViewModel?.deleteElement(element: element)
        }

        func saveEditingHistory() {
            commonViewModel?.saveEditingHistory()
        }

        func startObserving() {
            historyLoadStatus = nil
            editingHistory = nil
            showingElements = nil
            
            let commonViewModel = HistoryDetailCommonViewModel(historyId: historyId)
            self.commonViewModel = commonViewModel
            
            commonViewModel.historyLoadStatus.subscribe(scope: commonViewModel.viewModelScope) {
                self.historyLoadStatus = $0!
            }
            commonViewModel.editingHistory.subscribe(scope: commonViewModel.viewModelScope) {
                self.editingHistory = $0
            }
            commonViewModel.showingElements.subscribe(scope: commonViewModel.viewModelScope) {
                let showingElements: [HistoryElement]? = $0.map(Array.init)
                self.showingElements = showingElements
            }
        }

        func stopObserving() {
            commonViewModel?.dispose()
        }
    }
}

File: androidApp/src/main/java/com/example/stories/android/ui/historyDetail/HistoryDetailViewModel.kt
package com.example.stories.android.ui.historyDetail

import android.content.Context
import android.net.Uri
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import com.example.stories.android.util.ImageUtils
import com.example.stories.infrastructure.date.LocalDateRange
import com.example.stories.model.domain.model.HistoryElement
import com.example.stories.viewModel.HistoryDetailCommonViewModel
import kotlinx.coroutines.launch

class HistoryDetailViewModel(historyId: String) : ViewModel() {

    private val commonViewModel = HistoryDetailCommonViewModel(
        historyId = historyId,
        coroutineScope = viewModelScope,
    )

    val historyLoadStatus = commonViewModel.historyLoadStatus
    val editingHistory = commonViewModel.editingHistory

    fun enableEditMode() = commonViewModel.setEditMode()

    fun cancelEdit() = commonViewModel.cancelEdit()

    fun editItem(context: Context, element: HistoryElement, imageUri: Uri?) {
        viewModelScope.launch {
            val updatedElement = imageUri?.let {
                ImageUtils.uriToImageDomain(imageUri, context)
            }?.let { imageData ->
                (element as? HistoryElement.Image)?.setDataFromUrl(imageData)
            } ?: element
            commonViewModel.editElement(updatedElement)
        }
    }

    fun editTitle(newTitle: String) = commonViewModel.editTitle(newTitle)

    fun editDates(newDateRange: LocalDateRange) = commonViewModel.editDates(newDateRange)

    fun saveEditingHistory() = commonViewModel.saveEditingHistory()

    fun createTextElement(text: String) = commonViewModel.createTextElement(text)

    fun createImageElement(context: Context, imageUri: Uri) {
        viewModelScope.launch {
            val imageData = ImageUtils.uriToImageDomain(imageUri, context) ?: return@launch
            commonViewModel.createImageElement(imageData)
        }
    }

    fun swapElements(fromId: String, toId: String) = commonViewModel.swapElements(fromId, toId)

    fun deleteElement(element: HistoryElement) = commonViewModel.deleteElement(element)

    class Factory(private val historyId: String) : ViewModelProvider.NewInstanceFactory() {
        override fun <T : ViewModel> create(modelClass: Class<T>): T = HistoryDetailViewModel(historyId) as T
    }
}
